File: Dlib_CvExt/Painter.cpp
#include "stdafx.h"
#include "Painter.h"

using namespace Dlib::Basic;



NS_DLIB_CVEXT_BEGIN

Painter::Painter (Image& r_img, float scale, cv::Point2f offset, int shift) :
_r_img (r_img),
_scale (scale),
_offset (offset),
_shift (shift) {

}



Painter& Painter::line (cv::Point2f pt1, cv::Point2f pt2, Color color, int thickness) {

	color = color (_r_img.COLOR_SPACE ());
	cv::line (_r_img, _toShifted (pt1), _toShifted (pt2), color, thickness, cv::LINE_AA, _shift);
	return *this;

}



Painter& Painter::rectangle (cv::Point2f tl, cv::Point2f br, Color color, int thickness) {

	color = color (_r_img.COLOR_SPACE ());
	cv::rectangle (_r_img, _toShifted (tl), _toShifted (br), color, thickness, cv::LINE_AA, _shift);
	return *this;

}



Painter& Painter::rectangle (cv::Rect2f rect, Color color, int thickness) {

	color = color (_r_img.COLOR_SPACE ());
	rectangle (rect.tl (), rect.br (), color, thickness);
	return *this;

}



Painter& Painter::circle (cv::Point2f center, float radius, Color color, int thickness) {

	color = color (_r_img.COLOR_SPACE ());
	cv::circle (_r_img, _toShifted (center), _toShifted (radius), color, thickness, cv::LINE_AA, _shift);
	return *this;

}



Painter& Painter::ellipse (cv::Point2f center, cv::Size2f axes, double angle, double startAngle, double endAngle, Color color, int thickness) {

	color = color (_r_img.COLOR_SPACE ());
	cv::ellipse (_r_img, _toShifted (center), _toShifted (axes), angle, startAngle, endAngle, color, thickness, cv::LINE_AA, _shift);
	return *this;

}



Painter& Painter::polyline (Basic::ConstArrayRef<cv::Point2f> R_POINTS, Color color, int thickness, bool isConvexOrClosed) {

	color = color (_r_img.COLOR_SPACE ());
	if (thickness >= 0)
		cv::polylines (_r_img, std::vector<std::vector<cv::Point>> ({ _toShifted (R_POINTS) }), isConvexOrClosed, color, thickness, cv::LINE_AA, _shift);
	else if (isConvexOrClosed)
		cv::fillConvexPoly (_r_img, _toShifted (R_POINTS), color, cv::LINE_AA, _shift);
	else
		cv::fillPoly (_r_img, std::vector<std::vector<cv::Point>> ({ _toShifted (R_POINTS) }), color, cv::LINE_AA, _shift);
	return *this;

}


Painter& Painter::polylines (const std::vector<std::vector<cv::Point2f>>& R_POINTS_ARRAY, Color color, int thickness, bool isConvexOrClosed) {

	color = color (_r_img.COLOR_SPACE ());
	if (thickness >= 0)
		cv::polylines (_r_img, _toShifted (R_POINTS_ARRAY), isConvexOrClosed, color, thickness, cv::LINE_AA, _shift);
	else if (isConvexOrClosed)
		for (auto i = 0; i < R_POINTS_ARRAY.size (); ++i)
			cv::fillConvexPoly (_r_img, _toShifted (R_POINTS_ARRAY[i]), color, cv::LINE_AA, _shift);
	else
		cv::fillPoly (_r_img, _toShifted (R_POINTS_ARRAY), color, cv::LINE_AA, _shift);
	return *this;

}



Painter& Painter::marker (cv::Point2f pt, int type, int size, Color color, int thickness) {

	color = color (_r_img.COLOR_SPACE ());
	if (isnan (pt.x) || isnan (pt.y)) return *this;
	cv::drawMarker (_r_img, _transform (pt), color, type, size, thickness, cv::LINE_AA);
	return *this;

}



Painter& Painter::markers (ConstArrayRef<cv::Point2f> pts, int type, int size, Color color, int thickness) {

	color = color (_r_img.COLOR_SPACE ());
	for (int i = 0; i < pts.SIZE (); ++i)
		marker (pts[i], type, size, color, thickness);
	return *this;

}



Painter& Painter::fill (Color background) {

	background = background (_r_img.COLOR_SPACE ());
	cv::rectangle (_r_img, { 0, 0, _r_img.WIDTH (), _r_img.HEIGHT () }, background, -1);
	return *this;

}



Painter& Painter::text (const std::string& R_TEXT, cv::Point org, int fontFace, double fontScale, Color color, int thickness, bool bottomLeftOrigin) {

	color = color (_r_img.COLOR_SPACE ());
	cv::putText (_r_img, R_TEXT, (cv::Point)_transform (org), fontFace, fontScale, color, thickness, cv::LINE_AA, bottomLeftOrigin);
	return *this;

}



Painter& Painter::triangle (cv::Point2f pt, float r, float angle,
	Color color, int thickness) {

	color = color (_r_img.COLOR_SPACE ());
	std::vector<cv::Point2f> vertices (3);
	for (int i = 0; i < 3; ++i) {
		vertices[i].x = pt.x + r * std::cos ((angle + 120.0f * i) / 180.0f * 3.14159f);
		vertices[i].y = pt.y + r * std::sin ((angle + 120.0f * i) / 180.0f * 3.14159f);
	}
	polyline (vertices, color, thickness);
	return *this;

}

NS_DLIB_CVEXT_END

File: Runtime/Cv/Painter.cs
using System.Collections.Generic;
using Sirenix.OdinInspector;
using UDlib.Core;
using UDlib.Core.Image;
using UnityEditor;
using UnityEngine;

namespace UDlib.Cv {

	public class Painter {

		public enum Font {

			HERSHEY_SIMPLEX = 0, //!< normal size sans-serif font
			HERSHEY_PLAIN = 1, //!< small size sans-serif font
			HERSHEY_DUPLEX =
				2, //!< normal size sans-serif font (more complex than FONT_HERSHEY_SIMPLEX)
			HERSHEY_COMPLEX = 3, //!< normal size serif font
			HERSHEY_TRIPLEX =
				4, //!< normal size serif font (more complex than FONT_HERSHEY_COMPLEX)
			HERSHEY_COMPLEX_SMALL = 5, //!< smaller version of FONT_HERSHEY_COMPLEX
			HERSHEY_SCRIPT_SIMPLEX = 6, //!< hand-writing style font
			HERSHEY_SCRIPT_COMPLEX = 7, //!< more complex variant of FONT_HERSHEY_SCRIPT_SIMPLEX
			ITALIC = 16 //!< flag for italic font

		}

		public enum MarkerType {

			CROSS = 0, //!< A crosshair marker shape
			TILTED_CROSS = 1, //!< A 45 degree tilted crosshair marker shape
			STAR = 2, //!< A star marker shape, combination of cross and tilted cross
			DIAMOND = 3, //!< A diamond marker shape
			SQUARE = 4, //!< A square marker shape
			TRIANGLE_UP = 5, //!< An upwards pointing triangle marker shape
			TRIANGLE_DOWN = 6 //!< A downwards pointing triangle marker shape

		}

		private readonly TexImage _image;
		private readonly int _shift;

		public Painter (TexImage image, int shift = 2) {
			_image = image;
			_shift = shift;
		}

		public void Line (Vector2 pt1, Vector2 pt2, Color32 color, int thickness = 1) =>
			Plugin.drawLine (_image, pt1, pt2, color, thickness, _shift);

		public void Rectangle (Vector2 tl, Vector2 br, Color32 color, int thickness = 1) =>
			Plugin.drawRectangle (_image, tl, br, color, thickness, _shift);

		public void Rectangle (Rect rect, Color32 color, int thickness = 1) =>
			Plugin.drawRectangle (_image, rect.min, rect.max, color, thickness, _shift);

		public void Circle (Vector2 center, float radius, Color32 color, int thickness = 1) =>
			Plugin.drawCircle (_image, center, radius, color, thickness, _shift);

		public void Ellipse (Vector2 center, Vector2 axes, double angle, double startAngle,
			double endAngle, Color32 color, int thickness = 1) =>
			Plugin.drawEllipse (_image, center, axes, angle, startAngle, endAngle, color, thickness,
				_shift);

		public void Polyline (Array<Vector2> pts, Color32 color, int thickness = 1,
			bool isConvexOrClosed = true) =>
			Plugin.drawPolyline (_image, pts, color, thickness, isConvexOrClosed, _shift);

		public void Triangle (Vector2 pt, float r, float angle, Color32 color, int thickness = 1) =>
			Plugin.drawTriangle (_image, pt, r, angle, color, thickness, _shift);

		public void Marker (Vector2 pt, MarkerType type, int size, Color32 color,
			int thickness = 1) =>
			Plugin.drawMarker (_image, pt, type, size, color, thickness, _shift);

		public void Markers (Array<Vector2> pts, MarkerType type, int size, Color32 color,
			int thickness = 1) =>
			Plugin.drawMarkers (_image, pts, type, size, color, thickness, _shift);

		public void Fill (Color32 background) => Plugin.drawFill (_image, background, _shift);

		public void Text (string text, Vector2Int org, Font fontFace, double fontScale,
			Color32 color, int thickness = 1, bool bottomLeftOrigin = false) =>
			Plugin.drawText (_image, text, org, fontFace, fontScale, color, thickness,
				bottomLeftOrigin, _shift);

	}

}